static/js/tools/map/time_slider.tsx

```diff
@@ -0,0 +1,227 @@
+/**
+ * Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Time slider component.
+ */
+
+import React, { useEffect, useRef, useState } from "react";
+
+interface TimeSliderProps {
+  // Current date of map
+  currentDate: string;
+
+  // Selected dates to display on time slider
+  dates: Array<string>;
+
+  // Hash representing current time series
+  metahash: string;
+
+  // Whether the slider is enabled on refresh
+  // False if the map is displaying multiple dates (e.g. 'Best Available')
+  startEnabled: boolean;
+
+  // Fetches data for slider dates when play is pressed
+  onPlay(metahash: string, callback: () => void): any;
+
+  // Updates map date to slider date
+  updateDate(metahash: string, date: string): any;
+}
+
+export function TimeSlider(props: TimeSliderProps): JSX.Element {
+  const INTERVAL_MS = 500;
+  const SLIDER_MARGIN = 16;
+  const HANDLE_WIDTH = 4;
+  const HANDLE_MARGIN = SLIDER_MARGIN - HANDLE_WIDTH / 2;
+
+  const start = props.dates[0];
+  const end = props.dates[props.dates.length - 1];
+
+  const [currentDate, setCurrentDate] = useState(
+    props.startEnabled ? props.currentDate : "--"
+  );
+  const [enabled, setEnabled] = useState(props.startEnabled);
+  const [index, setIndex] = useState(
+    props.startEnabled ? getIndex(props.dates, props.currentDate) : -1
+  );
+  const [loaded, setLoaded] = useState(false);
+
+  // Number of pixels the handle is offset from the left edge of the slider bar
+  const [offset, setOffset] = useState(null);
+  const [play, setPlay] = useState(true);
+  const [timer, setTimer] = useState(null);
+
+  const firstUpdate = useRef(true);
+
+  useEffect(() => {
+    setLoaded(false);
+    setIndex(
+      props.startEnabled ? getIndex(props.dates, props.currentDate) : -1
+    );
+    setCurrentDate(props.startEnabled ? props.currentDate : "--");
+  }, [props.metahash]);
+
+  useEffect(() => {
+    if (enabled) {
+      setOffset(
+        getOffset(start, end, currentDate, SLIDER_MARGIN, HANDLE_MARGIN)
+      );
+    }
+  }, [currentDate, getOffset]);
+
+  useEffect(() => {
+    function handleResize() {
+      if (enabled) {
+        setOffset(
+          getOffset(start, end, currentDate, SLIDER_MARGIN, HANDLE_MARGIN)
+        );
+      }
+    }
+    window.addEventListener("resize", handleResize);
+  });
```

Comment:
I think you only want to add the eventListener the first time this component is created so you need `useEffect(() => {...}, []);`. Otherwise you will be adding the event listener on every re-render. 

Also, when adding event listeners, you want to remove the event listener once the component unmounts. so here, you want to add `return () => document.removeEventListener("resize", handleResize);` within the useEffect function

Reply:
so I added the removeEventListener to remove after each render, but if I add the [], it seems like it won't register state updates, so will just use the initial state (but the handleresize depends on the state) -- do you have any suggestions for this?

Reply:
ohh, if you need to watch for state updates, then this makes sense. Or if there's specific values in the state that you're watching for, you can add [] with those specific state variables